### DISCUSSION TITLE
JSONB Hybrid Properties

### DISCUSSION THREAD

**Discussion by MaartenUreel** [sqlalchemy/sqlalchemy#9660] [Usage Questions]
I have a model like this (I stripped a few properties to get to the essence):

```python
class CallFlowBlock(BaseModel):
    __tablename__ = "call_flow_blocks"

    data: Mapped[dict] = mapped_column(type_=JSON)

    def __repr__(self) -> str:
        return f"CallFlowBlock(id={self.id!r}, type={self.type_!r}, block_id={self.block_id!r}, description={self.description!r})"

    @hybrid_property
    def call_flow_id(self) -> int:
        return self.data["_call_flow"]

    @call_flow_id.expression
    @classmethod
    def call_flow_id(cls):
        return "data->>'_call_flow'"

    @hybrid_property
    def group_id(self) -> int:
        return self.data["_group"]

    @group_id.expression
    @classmethod
    def group_id(cls):
        return "data->>'_group'"
```

It's a bad database design, but I need to query this data with SQLAlchemy ORM now.

When writing this query however:

```python
    stmt = select(CallFlowBlock).where(CallFlowBlock.group_id == x.group.id, CallFlowBlock.call_flow_id == call_flow.id)
    x.logger.debug("Loading call flow blocks: %s", stmt)
```

The output is:
```sql
SELECT call_flow_blocks.data, call_flow_blocks.id 
FROM call_flow_blocks 
WHERE false
```

What is the correct way to define hybrid properties that map to keys on the JSONB object in the database? Do I need to wrap the literal outputs of my classmethods perhaps?

**Comment by zzzeek** [comment]
you would at the least wrap that SQL string in text() or literal_column() so that it's recognized as a SQL expression:

```
from sqlalchemy import literal_column

    @call_flow_id.expression
    @classmethod
    def call_flow_id(cls):
        return literal_column("data->>'_call_flow'")
```

however you will get much more mileage if you use the real [JSONB operator](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#sqlalchemy.dialects.postgresql.JSON)

```py
    @call_flow_id.expression
    @classmethod
    def call_flow_id(cls):
        return cls.data["_call_flow"].astext
```

**Reply by zzzeek** [reply]
you'd need to use postgresql.JSONB for that specific operator.  if using JSON, you can use [as_string](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.JSON.Comparator.as_string)

     cls.data["_call_flow"].as_string()

**Reply by MaartenUreel** [reply]
Is there an advantage in using postgresql.JSONB opposed to the generic JSON? What is best practice?

**Reply by zzzeek** [reply]
jsonb is considered to be more efficient and has more operators available:

https://www.dbvis.com/thetable/json-vs-jsonb-in-postgresql-a-complete-comparison/

but you lose database independence

**Reply by MaartenUreel** [reply]
Works, thanks for the great support.
